### PR TITLE
Commonize the profile selector component from Playbooks

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary

This PR is the first, small step towards commonizing the checklist components from Playbooks. That first step is commonzing the `ProfileSelector` component, which is made up of two smaller components.

So far, in its current state, this PR only commonizes the two smaller components `ProfileTag` and `ProfileButton`. The current next step is to gather feedback on those two components, and everything that went into making them common, before using those learnings to move on to `ProfileSelector`.

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: components/compass_theme_provider/compass_theme_provider.tsx [hunk 1]
```diff
@@ -5,7 +5,7 @@ import React, {useState, useEffect} from 'react';
 
 import ThemeProvider, {lightTheme} from '@mattermost/compass-components/utilities/theme';
 
-import {Theme} from 'mattermost-redux/types/themes';
+import {Theme} from '@mattermost/types/themes';
```
Comment (jwilander) [inline components/compass_theme_provider/compass_theme_provider.tsx:8]: Moved the `Theme` type into `types`

File: components/intl_provider/intl_provider.tsx [hunk 2]
```diff
@@ -7,7 +7,7 @@ import {IntlProvider as BaseIntlProvider} from 'react-intl';
 import {MessageFormatElement} from '@formatjs/icu-messageformat-parser';
 
 import {Client4} from 'mattermost-redux/client';
-import {setLocalizeFunction} from 'mattermost-redux/utils/i18n_utils';
+import {setLocalizeFunction} from '@mattermost/components/utils/i18n';
```
Comment (jwilander) [inline components/intl_provider/intl_provider.tsx:10]: Created a `utils` subfolder inside `components` and moved some of the needed mattermost-redux functions there

File: packages/components/src/constants/preferences.ts [hunk 3]
```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (jwilander) [inline packages/components/src/constants/preferences.ts:1]: Created a `constants` subfolder and moved all preferences over since these were needed by the components

File: packages/components/src/hooks/index.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (jwilander) [inline packages/components/src/hooks/index.ts:1]: Created a `hooks` subfolder that all hooks for components would be kept in - similar to what Playbooks does

File: packages/components/src/hooks/index.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import {getUser as fetchUser} from 'mattermost-redux/actions/entities/users';
+
+import {useEffect} from 'react';
+
+import {getPreferenceKey} from '../utils/preferences';
+import Preferences from '../constants/preferences';
+
+function getConfig(state: GlobalState): Partial<ClientConfig> {
+    return state.entities.general.config;
+}
```
Comment (jwilander) [inline packages/components/src/hooks/index.ts:20]: Re-implemented some mattermost-redux selectors here that were needed below. This isn't really ideal since now this code is in two places. I see two options:

1. Import it from mattermost-redux (bad because we want to reduce mattermost-redux usage)
2. Move selectors like this into `components` and export them to mattermost-redux which can re-export them to whatever is still using them. Then, eventually when mattermost-redux dies, everything needing these selectors can import them from `components` (not great, but better?)

Can anyone think of a different option?

File: packages/components/src/hooks/index.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+export function useUser(userId: string) {
+    const dispatch = useDispatch();
+    const user = useSelector((state) => getUser(state, userId));
+
+    useEffect(() => {
+        if (user) {
+            return;
+        }
+
+        dispatch(fetchUser(userId));
```
Comment (jwilander) [inline packages/components/src/hooks/index.ts:58]: I didn't want to but I needed to use the `fetchUser` action from mattermost-redux. Otherwise I would have had to move a ton of code over or re-implement it here. Not sure the best path forward here. Thoughts?

File: packages/components/src/profile/profile_tag/index.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+interface Props {
+    userId: string;
+}
+
+const ProfileTagConnector = (props: Props) => {
+    const user = useUser(props.userId);
+    const teamnameNameDisplaySetting = useTeammateNameDisplaySetting();
+
+    return (
+        <ProfileTag
```
Comment (jwilander) [inline packages/components/src/profile/profile_tag/index.tsx:19]: For components that need to be connected to redux, I put all redux connections in a higher order component and left the actual components ignorant of redux or where the props come from.

File: tests/helpers/themed-intl-test-helper.tsx [hunk 8]
```diff
@@ -4,7 +4,7 @@
 import React from 'react';
 
 import CompassThemeProvider from 'components/compass_theme_provider/compass_theme_provider';
-import {Theme} from 'mattermost-redux/types/themes';
+import {Theme} from '@mattermost/types/themes';
```
Comment (M-ZubairAhmed) [inline tests/helpers/themed-intl-test-helper.tsx:7]: These is conversation on how and where to move `theme` https://community.mattermost.com/core/pl/85auue3yytfutkts84rfefynpe

**Comment by M-ZubairAhmed** [comment]
Actually we are moving a lot of types from `packages/mattermost-redux/src/types` to other folder where it makes sense. 
https://mattermost.atlassian.net/browse/MM-46374
https://mattermost.atlassian.net/browse/MM-46373
https://mattermost.atlassian.net/browse/MM-46245
and others. I would like to propose to hold up until we merge those prs in.

File: packages/components/src/constants/preferences.ts (hunk 3 above)
Comment (hmhealey) [reply packages/components/src/constants/preferences.ts:1]: I'd rather not bring all of these constants over 1-1. Is there some subset this component needs?

File: packages/components/src/hooks/index.ts (hunk 5 above)
Comment (hmhealey) [reply packages/components/src/hooks/index.ts:20]: Does the component need to have these, or could we pass in the values from outside? It's not ideal since the users of this library will still have to get the user's profile and `teammateNameDisplay` preference themselves, but I'd prefer holding off on accessing the redux store from within this library as long as we can. I know that's likely a "when" rather than an "if", but I'd be worried that once we start doing it this way, it'll be hard to switch to something cleaner once that exists, so I'd like to hold off on that for as long as possible.

Something that also might work is having the web app provide these sort of hooks through Context. That way, components will know how to get a user, but they won't know where it comes from.

File: packages/components/package.json [hunk 9]
```diff
@@ -14,6 +14,9 @@
     "@babel/preset-env": "^7.16.11",
     "@babel/preset-react": "^7.16.7",
     "@babel/preset-typescript": "^7.16.7",
+    "@mattermost/client": "*",
```
Comment (hmhealey) [inline packages/components/package.json:17]: This versioning is something I actually don't know how it works. I don't know if we have to specify specific versions or if we use * and NPM figures out what versions are required for the package when it's shipped on NPM

File: packages/components/package.json [hunk 10]
```diff
@@ -14,6 +14,9 @@
...
+    "@mattermost/types": "*",
+    "mattermost-redux": "*",
```
Comment (hmhealey) [inline packages/components/package.json:19]: Based on my other comments, hopefully we don't actually need this

File: packages/components/src/hooks/index.ts (hunk 6 above)
Comment (hmhealey) [reply packages/components/src/hooks/index.ts:58]: I think we'd need to wait on the data manager before we could have a hook that fetches the user. Otherwise, we're risking making a ton more duplicate requests

File: packages/components/src/profile/profile_button.tsx [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        if (preferredName === userName) {
+            name = '@' + name;
+        }
+        return <span>{name}</span>;
+    };
+
+    return (
+        <Button
+            onClick={props.onClick}
+            className={props.profileButtonClass || 'PlaybookRunProfileButton'}
```
Comment (hmhealey) [inline packages/components/src/profile/profile_button.tsx:36]: There's some Playbooks stuff leaking into here

File: packages/components/src/profile/profile_tag/index.tsx (hunk 7 above)
Comment (hmhealey) [reply packages/components/src/profile/profile_tag/index.tsx:19]: Thanks for doing that, but I do think we'll want to leave that logic out of this package for now

File: packages/components/src/profile/profile_tag/profile_tag.tsx [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    if (user) {
+        const preferredName = displayUsername(user, teamnameNameDisplaySetting);
+        name = preferredName;
+        if (props.nameFormatter) {
+            name = props.nameFormatter(preferredName, user.username, user.first_name, user.last_name, user.nickname);
+        }
+    }
+
+    return (
+        <ProfileTagContainer className={classNames('ProfileTagContainer', props.classNames, props.className)}>
```
Comment (hmhealey) [inline packages/components/src/profile/profile_tag/profile_tag.tsx:79]: Does this need the `ProfileTagContainer` class name if its styled with styled-components? Also, we're duplicating the classNames prop here

File: packages/components/src/utils/users.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+    return '';
+}
+
+export function displayUsername(
+    user: UserProfile | null | undefined,
+    teammateNameDisplay: string,
+    useFallbackUsername = true,
+): string {
+    let name = useFallbackUsername ? localizeMessage('channel_loader.someone', 'Someone') : '';
```
Comment (hmhealey) [inline packages/components/src/utils/users.ts:27]: Instead of doing this in `displayUsername`, you could use `useIntl` in the component to do this translation. That'd also save you from having to make this package aware of how i18n works.

Plus, then we'd get rid of some usage of `localizeMessage`

File: packages/components/tsconfig.json [hunk 14]
```diff
@@ -13,9 +13,17 @@
         "noEmit": true,
         "declaration": true,
         "outDir": "dist",
+        "rootDir": "./src",
         "paths": {
             "mattermost-redux/*": ["./node_modules/mattermost-redux/src/*"],
-            "@mattermost/types/*": ["./node_modules/@mattermost/types/src/*"]
-        }
+            "@mattermost/types/*": ["./node_modules/@mattermost/types/src/*"],
```
Comment (hmhealey) [inline packages/components/tsconfig.json:19]: I think these paths entries might be able to be removed now since I added the parallel builders

File: packages/components/src/utils/users.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    if (user) {
+        if (teammateNameDisplay === Preferences.DISPLAY_PREFER_NICKNAME) {
```
Comment (hmhealey) [inline packages/components/src/utils/users.ts:29]: For these constants, maybe we just duplicate them in this file for now instead of trying to figure out where they should live now.

File: packages/components/package.json (hunk 9 above)
Comment (crspeller) [reply packages/components/package.json:17]: if we really do need this stuff we should set them up as workspaces.

File: packages/components/tsconfig.json [hunk 16]
```diff
@@ -1,7 +1,7 @@
 {
     "compilerOptions": {
         "target": "esnext",
-        "module": "esnext",
+        "module": "commonjs",
```
Comment (crspeller) [inline packages/components/tsconfig.json:4]: What's the reason for this change?

File: packages/components/src/hooks/index.ts (hunk 6 above)
Comment (crspeller) [reply packages/components/src/hooks/index.ts:58]: I would say that some of this is inevitable as we are moving components over. However the way it's done here is good because when we are ready to switch this out, it's all here.

File: packages/components/src/constants/preferences.ts (hunk 3 above)
Comment (crspeller) [reply packages/components/src/constants/preferences.ts:1]: Ya let's keep these local to where they are needed rather than in a constants file. I would like to kill that pattern.

Comment (jwilander) [reply packages/components/src/constants/preferences.ts:1]: What about constant that are needed across multiple components? The constants I needed here were used in these components but also in other places in the codebase. Having duplicates of the constants seems more problematic, doesn't it?

File: packages/components/src/hooks/index.ts (hunk 5 above)
Comment (jwilander) [reply packages/components/src/hooks/index.ts:20]: I do agree, I'd rather not have to bring actions and selectors in but I think the only way to pass them in through the outside would be to:

1. Not have any component hierarchy in the components library, and instead have "pure" (probably not the right word) components and then when they're used in a product, wrap them all with redux connector components and introduce the hierarchy there
2. Keep the hierarchy in the components library, but pass the props all the way down from the parent components at the top of the tree

I think I'd lean towards 1) otherwise we'll be doing a lot of props piping. Unless there's another option you can think of

File: packages/components/tsconfig.json (hunk 16 above)
Comment (jwilander) [reply packages/components/tsconfig.json:4]: Trying to get eslint to stop complaining about types

**Comment by jwilander** [comment]
Closing this for now since I haven't had time to get back to it and it's out of date.
